### DISCUSSION TITLE
F# support

### DISCUSSION THREAD

**Discussion by vzarytovskii** [badamczewski/PowerUp#7] [Ideas]
Hey @badamczewski, hope you're doing well.

Would you be interested in F# support (should be relatively easy, since C# already works)?

I would love to contribute.

**Comment by badamczewski** [comment]
Yes, there are plans to do it.

**Reply by badamczewski** [reply]
F# support has been added but only via Fsc.DLL. For a better experience, it would be good to interact with the compiler directly from C# and be able to Parse code and morph AST (As it's currently done in C#) to support some of the more advanced features like Benchmarks, Running Code, Shorthands in code, etc.

I haven't looked too much into this but there was no easy way to do it (AFAIK) but if you know a way, then just point me to it :)

**Reply by baronfel** [reply]
You'll want to grab the `FSharp.Compiler.Service` nuget package and look into either hand-rolling your own AST then feeding it to [FSharpChecker.Compile](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-codeanalysis-fsharpchecker.html#Compile), or using [FSharpChecker.ParseAndCheckFileInProject](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-codeanalysis-fsharpchecker.html#ParseAndCheckFileInProject) to get a valid AST, and then calling FSharpChecker.Compile with that.

The hard part will be creating an FSharpProjectOptions instance, but if you're in a very constrained runtime environment you can pretend you're an F# script and use the helper function [FSharpChecker.GetProjectOptionsFromScript](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-codeanalysis-fsharpchecker.html#GetProjectOptionsFromScript) to create one that works exactly like the `fsi` executable in the SDK you're running with.
